File: backend-dotnet/FinanceTracker.Api/Contracts/AccountDtos.cs
using System.ComponentModel.DataAnnotations;
using FinanceTracker.Api.Entities;
using FinanceTracker.Api.Infrastructure;

namespace FinanceTracker.Api.Contracts;

public sealed record AccountRequest(
    [param: NotBlank(ErrorMessage = "must not be blank")]
    string? name,
    [param: Required(ErrorMessage = "must not be null")]
    AccountType? type,
    [param: Required(ErrorMessage = "must not be null")]
    [param: MinValue("0.0", ErrorMessage = "must be greater than or equal to 0.0")]
    decimal? openingBalance,
    string? institutionName
);

public sealed record AccountResponse(
    Guid id,
    string name,
    AccountType type,
    decimal openingBalance,
    decimal currentBalance,
    string? institutionName,
    DateTimeOffset createdAt
);

public sealed record TransferRequest(
    [param: Required(ErrorMessage = "must not be null")]
    Guid? sourceAccountId,
    [param: Required(ErrorMessage = "must not be null")]
    Guid? targetAccountId,
    [param: Required(ErrorMessage = "must not be null")]
    [param: MinValue("0.01", ErrorMessage = "must be greater than or equal to 0.01")]
    decimal? amount,
    [param: Required(ErrorMessage = "must not be null")]
    DateOnly? transactionDate,
    string? note
);

File: backend/src/main/java/com/finance/dto/AccountDtos.java
package com.finance.dto;

import com.finance.entities.AccountType;
import jakarta.validation.constraints.DecimalMin;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import java.math.BigDecimal;
import java.time.OffsetDateTime;
import java.util.UUID;

public class AccountDtos {

    public record AccountRequest(
            @NotBlank String name,
            @NotNull AccountType type,
            @NotNull @DecimalMin(value = "0.0", inclusive = true) BigDecimal openingBalance,
            String institutionName
    ) {
    }

    public record AccountResponse(
            UUID id,
            String name,
            AccountType type,
            BigDecimal openingBalance,
            BigDecimal currentBalance,
            String institutionName,
            OffsetDateTime createdAt
    ) {
    }

    public record TransferRequest(
            @NotNull UUID sourceAccountId,
            @NotNull UUID targetAccountId,
            @NotNull @DecimalMin(value = "0.01") BigDecimal amount,
            @NotNull java.time.LocalDate transactionDate,
            String note
    ) {
    }
}
